cranelift/codegen/src/isa/riscv64/inst.isle

```diff
@@ -999,50 +988,121 @@
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rr_imm12 (AluOPRRI.Andi) val (imm12_const 255))))
+
+;; When signed extending from 32 to 64 bits we can use a
+;; `addiw val 0`. Also known as a `sext.w`
+(rule 2 (extend val (ExtendOp.Signed) $I32 $I64)
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rr_imm12 (AluOPRRI.Addiw) val (imm12_const 0))))
+
+
+;; No point in trying to use `packh` here to zero extend 8 bit values
+;; since we can just use `andi` instead which is part of the base ISA.
+
+;; If we have the `zbkb` extension `packw` can be used to zero extend 16 bit values
+(rule 3 (extend val (ExtendOp.Zero) $I16 (fits_in_64 _))
+  (if-let $true (has_zbkb))
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rrr (AluOPRRR.Packw) val (zero_reg))))
 
+;; If we have the `zbkb` extension `pack` can be used to zero extend 32 bit registers
+(rule 4 (extend val (ExtendOp.Zero) $I32 $I64)
+  (if-let $true (has_zbkb))
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rrr (AluOPRRR.Pack) val (zero_reg))))
 
-;;;; for I128 unsigned extend.
-(rule 1
-  (lower_extend r $false 64 128)
-  (value_regs (gen_move2 r $I64 $I64) (load_u64_constant 0)))
 
-(rule
-  (lower_extend r $false from_bits 128)
-  (value_regs (gen_extend r $false from_bits 64) (load_u64_constant 0)))
+;; If we have the `zbb` extension we can use the dedicated `sext.b` instruction.
+(rule 5 (extend val (ExtendOp.Signed) $I8 (fits_in_64 _))
+  (if-let $true (has_zbb))
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rr_imm12 (AluOPRRI.Sextb) val (imm12_const 0))))
+
+;; If we have the `zbb` extension we can use the dedicated `sext.h` instruction.
+(rule 6 (extend val (ExtendOp.Signed) $I16 (fits_in_64 _))
+  (if-let $true (has_zbb))
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rr_imm12 (AluOPRRI.Sexth) val (imm12_const 0))))
+
+;; If we have the `zbb` extension we can use the dedicated `zext.h` instruction.
+(rule 7 (extend val (ExtendOp.Zero) $I16 (fits_in_64 _))
+  (if-let $true (has_zbb))
+  (let ((val Reg (value_regs_get val 0)))
+    (alu_rr_imm12 (AluOPRRI.Zexth) val (imm12_const 0))))
+
+;;; Signed rules extending to I128
+;; Extend the bottom part, and extract the sign bit from the bottom as the top
+(rule 8 (extend val (ExtendOp.Signed) (fits_in_64 from_ty) $I128)
+  (let ((val Reg (value_regs_get val 0))
+        (low Reg (extend val (ExtendOp.Signed) from_ty $I64))
+        (high Reg (alu_rr_imm12 (AluOPRRI.Srai) low (imm12_const 63))))
+    (value_regs low high)))
+
+
+;;; Unsigned rules extending to I128
+;; Extend the bottom register to I64 and then just zero out the top half.
+(rule 9 (extend val (ExtendOp.Zero) (fits_in_32 from_ty) $I128)
+  (let ((val Reg (value_regs_get val 0))
+        (extended Reg (extend val (ExtendOp.Zero) from_ty $I64)))
+    (value_regs extended (load_u64_constant 0))))
+
+(rule 10 (extend val (ExtendOp.Zero) $I64 $I128)
+  (value_regs
+    (value_regs_get val 0)
+    (load_u64_constant 0)))
```

Comment:
I think that ended up being a bit cleaner! I like having the signed and unsigned rules being "visually" equal. Thanks!

> They do currently need a different priority than all the other rules though since ISLE can't yet figure out that $I128 doesn't overlap with fits_in_64. I want to fix that sooner or later...

Oh, I think that might be why I started thinking that ISLE wouldn't reason past `fits_in_64`, good to know that its only in I128's.

Reply:
That's not quite it. The issue is that ISLE doesn't know anything about when an extractor like `fits_in_64` will match. If you have two patterns where one is `$I64` and the other is `$I32`, it can see that those are different. If the patterns are both `fits_in_64`, it knows that both patterns will produce the same result on the same input so it can check deeper in the tree of patterns. But if the patterns are `fits_in_64` versus `fits_in_32`, or versus `$I16`, ISLE can't draw any conclusions about whether those overlap or not.

I want to fix this by making ISLE understand or-patterns, then defining `fits_in_64` as the equivalent of `$I8 | $I16 | ...`. By making more semantics visible to ISLE, we'll get better pattern-matching code out as well as more precise overlap checks. Someday...

Reply:
Yeah, makes sense, Thanks for explaining that!